lib/components/user/notification-prefs-pane.js

```diff
@@ -28,85 +36,393 @@ const allowedNotificationChannels = [
+                    />
+                  )
+                }
+              }
+            </Formik>
+          )}
+        </Details>
+      </div>
+    )
+  }
+}
+
+export default NotificationPrefsPane
+
+/**
+ * Sub-component that handles phone number and validation code editing and validation intricacies.
+ */
+class PhoneNumberEditor extends Component {
+  constructor (props) {
+    super(props)
+
+    const { initialPhoneNumber, initialPhoneNumberVerified } = props
+    this.state = {
+      // Holds the last known verified phone number.
+      currentNumber: initialPhoneNumber,
+
+      // The verified status of the above number.
+      currentNumberVerified: initialPhoneNumberVerified,
+
+      // If true, phone number is being edited.
+      // For new users, render component in editing state.
+      isEditing: isBlank(initialPhoneNumber),
+
+      // If true, a phone verification request has been sent and the UI is awaiting the user sending the code.
+      isPending: false,
+
+      // Holds the new phone number (+15555550123 format) entered by the user
+      // (outside of Formik because (Phone)Input does not have a standard onChange event or simple valitity test).
+      newPhoneNumber: ''
+    }
+  }
+
+  _handleEditNumber = () => {
+    this.setState({
+      isEditing: true,
+      isPending: false,
+      newPhoneNumber: ''
+    })
+  }
+
+  _handleNewPhoneNumberChange = newPhoneNumber => {
+    this.setState({
+      newPhoneNumber
+    })
+  }
+
+  _handleCancelEditNumber = () => {
+    this.setState({
+      isEditing: false,
+      isPending: false
+    })
+  }
+
+  _handlePhoneNumberKeyDown = e => {
+    if (e.keyCode === 13) {
+      // On the user pressing enter (keyCode 13) on the phone number field,
+      // prevent form submission and request the code.
+      e.preventDefault()
+      this._handleRequestCode()
+    }
+  }
+
+  _handleValidationCodeKeyDown = e => {
+    if (e.keyCode === 13) {
+      // On the user pressing enter (keyCode 13) on the validation code field,
+      // prevent form submission and send the validation code.
+      e.preventDefault()
+      this._handleSubmitCode()
+    }
+  }
+
+  _handleRequestCode = async () => {
+    // Send phone verification request with the entered values.
+    const { onRequestCode, setFieldValue, setFieldTouched } = this.props
+    const { currentNumber, currentNumberVerified, newPhoneNumber } = this.state
+
+    // If user enters the same phone number as their current verified number,
+    // just cancel editing without sending a verification SMS.
+    if (newPhoneNumber === currentNumber && currentNumberVerified) {
+      this._handleCancelEditNumber()
+    } else {
+      // Empty the input box (and disable the Send text button).
+      this.setState({
+        newPhoneNumber: ''
+      })
+
+      const { isPhoneNumberVerified, phoneNumber } = await onRequestCode(newPhoneNumber)
+
+      if (phoneNumber === newPhoneNumber && !isPhoneNumberVerified) {
```

Comment:
This again has a lot of code that waits for application state to change and relies on the component's props to become updated. This is somewhat confusing and I think would be better handled by splitting this logic in 2. Move the code requesting a code into an action. For the rest of this code, create a `componentDidUpdate` method where you can read in the newly updated props and call `setFieldValue` and `setFieldTouched` appropriately.